### DISCUSSION TITLE
Include CommunityID::hash_conn in base

### DISCUSSION THREAD

**Discussion by awelzel** [zeek/zeek#2542] [Ideas]
(Mostly food of thought, I was writing docs and wanted to use some package as an example and community-id came to mind, but then it's a plugin and needs a development setup to try).

CommunityID has established itself and is implemented by quite a few products/projects:
https://github.com/corelight/community-id-spec#production-implementations

Zeek does not support CommunityID out of the box. Adding it to a vanilla installation involves setting up a development environment to compile an external plugin and the friction that may come with this. Similar arguments about convenience can be made as for AF_Packet though that was certainly more of a "applies to everyone on Linux".

Still, could minimally the CommunityID::hash_conn bif be included in-tree? It's a hash function specifically tailored for network monitoring applications and yet Zeek base does not offer support for it. This could allow the community-id package to become script-only in the future, too.

EDIT: Seems there are some script-level options involved that the bif is using, so it's not just the bif.

**Comment by ckreibich** [comment]
I'm torn here — in terms of complexity it's no big deal to include it, but it's also been a nice example of modular functionality. Not sure how many people see these discussions, but it'd be great to hear feedback on this from Community ID users to understand how much of a pain this installation has been for folks in practice.

At first I thought we could expose the hashing API to the script layer and reduce the Community ID package to pure scripts, but that's tricky because the hash inputs need some low-level byte wrangling (see the BiF for details). So I don't think that's a good way forward.

**Reply by grigorescu** [reply]
> At first I thought we could expose the hashing API to the script layer and reduce the Community ID package to pure scripts, but that's tricky because the hash inputs need some low-level byte wrangling (see the BiF for details).

I'd like to see the necessary BIFs just added, regardless of the decision here. There are times when it'd be useful to be able to do bit-wrangling in scripts (an issue I ran into long ago was trying to implement a DGA for some specific piece of malware, and was unable to).

It kind of feels like we have two quasi-types in bytestring and hexstring. We have helper functions like `bytestring_to_count`, and we can convert between bytestring and hexstring with `bytestring_to_hexstring` and `hexstring_to_bytestring`. But I wasn't able to find a way to go from a count, float, etc. to a bytestring.

Confusing matters is the fact that we also have `raw_bytes_to_v4_addr` and `raw_bytes_to_v6_addr`, which is the same as a "bytestring," but is just called differently.

`bytestring_to_count` is the only function that lets the user specify endian-ness.

I think the minimum we would need would be:

* Add a `count_to_bytestring` and an `addr_to_bytestring`
* Add a `bytestring_to_addr`, which combines `raw_bytes_to_v4_addr` and `raw_bytes_to_v6_addr`, and determines it automatically based on length. and which allows the user to specify endian-ness.

I think that should give us everything we need to generate the data, and feed it into the SHA1 file analyzer and generate the correct hash.

**Comment by 0xxon** [comment]
I assume that this is not a problem, because we have a lot of features that already require a development environment on the system. The main one here is Spicy - we will start including spicy parsers in Zeek directly, and I assume spicy parsers in general will become very common. They require a dev-environment. Compiling Zeek scripts does too.

**Comment by timwoj** [comment]
This work was completed, and CommunityID will part of the base installation as of Zeek 6.0.
